### DISCUSSION TITLE
How do I use Delve with bubbletea?

### DISCUSSION THREAD

**Discussion by er4hn** [charmbracelet/bubbletea#241] [Q&A] [answered]
Hello,

I'm trying to debug an issue with my code and I would like to use Delve (dlv) to do so. I have a basic bubbletea program which I start with bubbletea.NewProgram. When I start it under dlv (i.e. dlv debug -> continue) the continue loop seems stuck and I do not see any output. Should I be doing something special to use Delve?

Here it my set of goroutines in case something is blocked there:

(dlv) goroutines
* Goroutine 1 - User: /usr/local/go/src/syscall/asm_linux_amd64.s:26 syscall.Syscall (0x4a975b) (thread 38764) [chan receive]
  Goroutine 2 - User: /usr/local/go/src/runtime/proc.go:367 runtime.gopark (0x438672) [force gc (idle)]
  Goroutine 3 - User: /usr/local/go/src/runtime/proc.go:367 runtime.gopark (0x438672) [GC sweep wait]
  Goroutine 4 - User: /usr/local/go/src/runtime/proc.go:367 runtime.gopark (0x438672) [GC scavenge wait]
  Goroutine 17 - User: /usr/local/go/src/runtime/proc.go:367 runtime.gopark (0x438672) [finalizer wait]
  Goroutine 18 - User: /home/erahn/go/pkg/mod/github.com/charmbracelet/bubbletea@v0.20.0/tea.go:347 github.com/charmbracelet/bubbletea.(*Program).StartReturningModel.func2 (0x52d0a5) [select]
  Goroutine 33 - User: /usr/local/go/src/runtime/proc.go:367 runtime.gopark (0x438672) [select]
  Goroutine 49 - User: /usr/local/go/src/runtime/sigqueue.go:169 os/signal.signal_recv (0x4621f8) (thread 38791)
[8 goroutines]

**Comment by muesli** [comment, marked as answer]
Since BubbleTea apps are in control of stdin/stdout, you need to start it in a headless delve debugger, and then connect to it:

```bash
# Start the debugger:
$ dlv debug --headless .
API server listening at: 127.0.0.1:34241

# Connect to it from another terminal:
$ dlv connect 127.0.0.1:34241
```

Note that the default port used will vary on your system and per run, so actually watch out what address the first dlv run tells you to connect to!

**Reply by er4hn** [reply]
Thank you! I'm now able to get a lot more insights into where I am going wrong 🔬
